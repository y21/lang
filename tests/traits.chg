type Foo<T> = { x: T };

trait Default {
    fn default(): Self;
    fn method(self);
}
impl Default for i32 {
    fn default(): Self {
        return 0;
    }
    fn method(self) {}
}

trait Bar {
    fn baz<U>(self, u: U): U;
}
impl<T> Bar for Foo<T> {
    fn baz<U>(self, u: U): U {
        return u;
    }
}

fn main(): i32 {
    let x: i32 = 10;
    x.method();
    return Default::default();
    let b: Foo<i32> = .{ x: 10 };
    return b.baz(55);
    let x: Foo<i32> = .{ x: Default::default() };
    print(x);
}

fn print<T>(x: T) where T: Bar {}
