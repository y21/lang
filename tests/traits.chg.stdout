expr @ 9:16 9:17 => i32
expr @ 9:9 9:17 => never
expr @ 8:24 11:0 => ()
expr @ 11:21 12:0 => ()
expr @ 19:16 19:17 => undefined
expr @ 19:9 19:17 => never
expr @ 18:30 21:0 => ()
expr @ 24:18 24:20 => i32
expr @ 25:5 25:6 => i32
expr @ 25:5 25:15 => ()
expr @ 26:21 26:28 => fn default(...)
expr @ 26:21 26:30 => i32
expr @ 26:5 26:30 => never
expr @ 27:29 27:31 => i32
expr @ 27:23 27:33 => {x: i32}
expr @ 28:12 28:13 => Foo<i32>
expr @ 28:18 28:20 => i32
expr @ 28:12 28:21 => i32
expr @ 28:5 28:21 => never
expr @ 29:38 29:45 => fn default(...)
expr @ 29:38 29:47 => i32
expr @ 29:23 29:49 => {x: i32}
expr @ 30:5 30:10 => fn print(...)
expr @ 30:11 30:12 => Foo<i32>
expr @ 30:5 30:13 => ()
expr @ 23:16 32:0 => ()
expr @ 33:32 34:1 => ()



define void @root$i32$$LB$0$RB$method$LT$i32$GT$(i32 %lt.0) {
prologue:
%l.0 = alloca i32
store i32 %lt.0, i32* %l.0
br label %bb.0

bb.0:
ret void

}

define i32 @root$i32$$LB$0$RB$default$LT$i32$GT$() {
prologue:
%l.0 = alloca i32
br label %bb.0

bb.0:
store i32 0, i32* %l.0
%ret.0 = load i32, i32* %l.0
ret i32 %ret.0

}

define i32 @root$Foo$$LB$3$RB$baz$LT$root$Foo$LT$i32$RT$$$i32$$i32$GT$({i32} %lt.1, i32 %lt.2) {
prologue:
%l.0 = alloca i32
%l.1 = alloca {i32}
%l.2 = alloca i32
store {i32} %lt.1, {i32}* %l.1
store i32 %lt.2, i32* %l.2
br label %bb.0

bb.0:
%l.3 = load i32, i32* %l.2
store i32 %l.3, i32* %l.0
%ret.0 = load i32, i32* %l.0
ret i32 %ret.0

}

define void @root$print$LT$root$Foo$LT$i32$RT$$GT$({i32} %lt.0) {
prologue:
%l.0 = alloca {i32}
store {i32} %lt.0, {i32}* %l.0
br label %bb.0

bb.0:
ret void

}

define i32 @main() {
prologue:
%l.0 = alloca i32
%l.1 = alloca i32
%l.5 = alloca {i32}
%l.8 = alloca {i32}
br label %bb.0

bb.0:
store i32 10, i32* %l.1
%l.2 = load i32, i32* %l.1
%l.3 = call {} bitcast(void(i32)* @root$i32$$LB$0$RB$method$LT$i32$GT$ to {}(i32)*)(i32 %l.2)
br label %bb.1

bb.1:
%l.4 = call i32 @root$i32$$LB$0$RB$default$LT$i32$GT$()
br label %bb.2

bb.2:
store i32 %l.4, i32* %l.0
%ret.0 = load i32, i32* %l.0
ret i32 %ret.0

bb.3:
%t.1 = alloca {i32}
%p.2 = getelementptr {i32}, {i32}* %t.1, i32 0, i32 0
store i32 10, i32* %p.2
%t.3 = load {i32}, {i32}* %t.1
store {i32} %t.3, {i32}* %l.5
%l.6 = load {i32}, {i32}* %l.5
%l.7 = call i32 @root$Foo$$LB$3$RB$baz$LT$root$Foo$LT$i32$RT$$$i32$$i32$GT$({i32} %l.6, i32 55)
br label %bb.4

bb.4:
store i32 %l.7, i32* %l.0
%ret.4 = load i32, i32* %l.0
ret i32 %ret.4

bb.5:
%l.9 = call i32 @root$i32$$LB$0$RB$default$LT$i32$GT$()
br label %bb.6

bb.6:
%t.5 = alloca {i32}
%p.6 = getelementptr {i32}, {i32}* %t.5, i32 0, i32 0
store i32 %l.9, i32* %p.6
%t.7 = load {i32}, {i32}* %t.5
store {i32} %t.7, {i32}* %l.8
%l.11 = load {i32}, {i32}* %l.8
%l.10 = call {} bitcast(void({i32})* @root$print$LT$root$Foo$LT$i32$RT$$GT$ to {}({i32})*)({i32} %l.11)
br label %bb.7

}


